### PR TITLE
Buy web UI

### PR DESCRIPTION
**Pull request by Pavneet-Sing**
Resolves https://github.com/brave/brave-browser/issues/31494

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

[buy-webui-2.webm](https://github.com/brave/brave-core/assets/11755381/399efa73-3a95-4752-96c9-63a7db76f788)

## Additional Info
Will run all CIs after some approvals though LMK if it is required.

### CODE REVIEW

File: browser/ui/webui/brave_web_ui_controller_factory.cc [hunk 1]
```diff
@@ -199,7 +200,8 @@ WebUIFactoryFunction GetWebUIFactoryFunction(WebUI* web_ui,
 #if BUILDFLAG(IS_ANDROID)
       (url.is_valid() && url.host_piece() == kWalletPageHost &&
        (url.path() == kWalletSwapPagePath ||
-        url.path() == kWalletSendPagePath)) ||
+        url.path() == kWalletSendPagePath ||
+        url.path() == kWalletBuyPagePath)) ||
```
Comment (Pavneet-Sing) [inline browser/ui/webui/brave_web_ui_controller_factory.cc:204]: info: using buy to match BSS though LMK if I should use `kWalletFundPagePath`

File: components/brave_wallet_ui/page/screens/fund-wallet/fund-wallet.tsx [hunk 2]
```diff
@@ -395,7 +401,7 @@ export const FundWalletScreen = () => {
                 onSubmit={nextStep}
                 disabled={!isNextStepEnabled}
                 isV2={true}
-                minWidth='436px'
+                minWidth='360px'
```
Comment (Pavneet-Sing) [inline components/brave_wallet_ui/page/screens/fund-wallet/fund-wallet.tsx:404]: Info: Updated for Android and [the size in design spec is (324px)](https://www.figma.com/file/muVcmcu4RFNu0LkXe9VRDl/Wallet-2.0?type=design&node-id=3390-259364&mode=dev).

File: components/brave_wallet_ui/common/hooks/use-multi-chain-buy-assets.ts [hunk 3]
```diff
@@ -155,13 +155,18 @@ export const useMultiChainBuyAssets = () => {
...
+          chrome.tabs.create({ url }, () => {
+            if (chrome.runtime.lastError) {
+              console.error(
+                'tabs.create failed: ' + chrome.runtime.lastError.message
+              )
+            }
+          })
+        } else {
+          // Tabs.create is desktop specific. Using window.open for android.
+          window.open(url, "_blank");
```
[the remark this answers is not in the document]
Comment (Pavneet-Sing) [reply components/brave_wallet_ui/common/hooks/use-multi-chain-buy-assets.ts]: Addressed : https://github.com/brave/brave-core/pull/19334/commits/a2f1b788d350f58db9db7feb86f481114e3e82f9

**Review by Douglashdaniel — APPROVED** [review]
++ Front-end

**Review by mkarolin — APPROVED** [review]
`GRD`++

File: components/brave_wallet_ui/common/hooks/use-multi-chain-buy-assets.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (thypon) [reply components/brave_wallet_ui/common/hooks/use-multi-chain-buy-assets.ts]: If you don't comment and wait the `security-action` to run again, the comment is automatically removed, btw :)

**Review by simoarpe — APPROVED** [review]
👍 Android looks good

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java [hunk 4]
```diff
@@ -237,40 +236,28 @@ public static void hideKeyboard(Activity activity) {
...
                 break;
             default:
                 throw new IllegalStateException(
                         String.format("Activity not found for type %s.", activityType));
         }
+        try {
+            BraveActivity.getBraveActivity().openNewOrSelectExistingTab(webWalletUrl, true);
+            TabUtils.bringChromeTabbedActivityToTheTop(activity);
+        } catch (BraveActivity.BraveActivityNotFoundException e) {
+            Log.e(TAG, "Error while opening swap tab.", e);
```
Comment (vadimstruts) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java:259]: nit: Looks like the log description is related not only to swap tab.

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java:259]:
```suggestion
            Log.e(TAG, "Error while opening wallet tab.", e);
```

**Review by vadimstruts — APPROVED** [review]
++

**Comment by simoarpe** [comment]
Lots of conflicts, because of https://github.com/brave/brave-core/pull/19256
If you need help feel free to ping me @Pavneet-Sing

**Comment by Pavneet-Sing** [comment]
> Lots of conflicts, because of #19256 If you need help feel free to ping me @Pavneet-Sing

Thanks though no problem at all. They have been already resolved.

[buy-webui-2.webm](https://github.com/brave/brave-core/assets/11755381/847e5e6d-bd1c-49da-9920-ac65443da195)

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Pull request merged**